tortoisebeacon/tortoise_beacon.go

```diff
@@ -460,12 +459,39 @@ func (tb *TortoiseBeacon) proposalPhaseImpl(ctx context.Context, epoch types.Epo
 	}
 
 	logger.With().Info("sent proposal", log.String("message", m.String()))
+	return nil
+}
 
-	tb.mu.Lock()
-	defer tb.mu.Unlock()
-
-	tb.incomingProposals.valid = append(tb.incomingProposals.valid, proposedSignature)
+func (tb *TortoiseBeacon) getProposalChannel(ctx context.Context, epoch types.EpochID) chan *proposalMessageWithReceiptData {
+	tb.mu.RLock()
+	defer tb.mu.RUnlock()
 
+	if epoch < tb.epochInProgress {
+		tb.logger.WithContext(ctx).With().Debug("proposal too old, do not accept",
+			log.Uint32("proposal_epoch", uint32(epoch)))
+		return nil
+	}
+	if epoch == tb.epochInProgress {
+		ongoing := tb.proposalPhaseFinishedTime == time.Time{}
+		if !ongoing {
+			tb.logger.WithContext(ctx).With().Debug("proposal phase ended, do not accept",
+				log.Uint32("proposal_epoch", uint32(epoch)))
+			return nil
+		}
+		return tb.getOrCreateProposalChannel(epoch)
+	}
+	// always accept proposals for the next epoch, but not too far in the future
+	if epoch == tb.epochInProgress+1 {
+		ch := tb.getOrCreateProposalChannel(epoch)
+		if len(ch) == proposalChanCapacity {
+			// the reader loop is not started for the next epoch yet. drop old messages if it's already full
+			msg := <-ch
```

Comment:
i think it is correct to select with default to make in nonblocking. receive is not synchronous with length check, so it may block in some testing for example

Reply:
nice catch. thanks. done